File: NAQ2017/C.cpp
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <algorithm>
#include <cmath>
#include <climits>
#include <limits>
#include <vector>
#include <queue>
#include <cstring>
#include <iterator>
#include <list>
#include <set>
#include <map>
#include <bitset>
using namespace std;

typedef long long LL;

LL n;
LL coins[110];
LL greedy(LL x) {
    LL y = 0;
    for (int i = n-1; i >= 0; i--) {
        y += x / coins[i];
        x %= coins[i];
    }
    return y;
}

LL dp[2000010];
int main()
{
    cin >> n;
    for (int i = 0; i < n; i++)
        cin >> coins[i];

    LL N = coins[n-1]+coins[n-2];
    //LL infinity = 10000000000LL;
    LL infinity = n*N;
    for (int i = 0; i < N; i++)
        dp[i] = infinity;
    dp[0] = 0;
    for (int i = 0; i < n; i++) {
        LL c = coins[i];
        for (LL j = c; j < N; j++)
            dp[j] = min(dp[j],dp[j-c]+1);
    }

    /*
    for (int i = 0; i < N; i++) {
        cout << dp[i] << " ";
    }
    cout << endl;
    */

    for (int i = 0; i < N; i++)
        if (greedy(i) != dp[i]) {
            /*
            cout << "counterexample is " << i << endl;
            cout << "greedy is " << greedy(i) << endl;
            cout << "dp is " << dp[i] << endl;
            */
            cout << "non-canonical" << endl;
            return 0;
        }

    cout << "canonical" << endl;
    return 0;
}

File: NAQ2017/C.py
import sys
n = int(input())
coins = [int(a) for a in input().split()]

#print(coins)
def greedy(x):
    y = 0
    for c in reversed(coins):
        y += x // c
        x %= c
    return y

N = coins[-1]+coins[-2]
infinity = 10**10
dp = [infinity] * N
greedy = [infinity] * N

dp[0] = 0
for c in coins:
    for i in range(c,N):
        dp[i] = min(dp[i],dp[i-c]+1)

for i in range(N):
    if greedy(i) != dp[i]:
        #print('counter example:', i)
        #print('greedy:', greedy(i))
        #print('dp:', dp[i])
        print('non-canonical')
        sys.exit()
print('canonical')
